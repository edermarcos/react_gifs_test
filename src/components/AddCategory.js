import React from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategory }) => {
  const [inputValue, setInputValue] = React.useState('');

  const handleSubmit = (e) => {
    e.preventDefault();

    if (!inputValue.trim()) {
      return;
    }

    /*
     * Se recibe como parametro el setter desde el padre
     * Por default, entites es el parametro que retorna y se crea un nuevo arreglo
     * */
    setCategory(entities => [inputValue, ...entities]);
    setInputValue('');
  };

  return (
    <form onSubmit={ handleSubmit }>
      <p>{ inputValue }</p>
      <input type="text"
             value={ inputValue }
             onChange={ (event) => setInputValue(event.target.value) }/>
    </form>
  );
};

AddCategory.propTypes = {
  setCategory: PropTypes.func.isRequired
};
