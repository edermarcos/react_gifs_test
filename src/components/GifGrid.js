import React from 'react';
import PropTypes from 'prop-types';

import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category, deleteCategoryById }) => {
  const { data: images, loading } = useFetchGifs(category);

  return (
    <div>
      <div className="header">
        <h3>Category: { category }</h3>

        <div>
          <button onClick={ () => deleteCategoryById(category) }>Delete</button>
        </div>
      </div>

      { loading && <p>Loading...</p> }

      <div className="card-grid">
        {
          images.map(entity => (
            <GifGridItem key={ entity.id }
                         { ...entity }
            />
          ))
        }
      </div>
    </div>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
  deleteCategoryById: PropTypes.func,
};
