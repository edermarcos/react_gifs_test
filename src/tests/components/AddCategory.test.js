import React from 'react';
import { shallow } from 'enzyme';

import { AddCategory } from '../../components/AddCategory';

describe('Tests for <AddCategory />', () => {
  const setCategory = jest.fn();
  let wrapper;

  beforeEach(() => {
    /*
     * Limpia un mock o simulacion de algun evento
     * */
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategory={ setCategory }/>);
  });

  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de cambiar la caja del texto', () => {
    const input = wrapper.find('input');
    const value = 'Hola Mundo';

    /*
     * Simula el evento cuando se hace el onChange
     * Event es para simular el event.target.value del html
     * */
    const event = { target: { value } };
    input.simulate('change', event);
    const p = wrapper.find('p').text().trim();
    expect(p).toBe(value);
  });

  test('should not sent the form with the submit', () => {
    /*
     * Find the form an simulate the submit event
     * */
    wrapper.find('form').simulate('submit', { preventDefault() {} });

    /*
     * Evalua que la funcion no deberia ser llamada
     * */
    expect(setCategory).not.toHaveBeenCalled();
  });

  test('Debe de llamar a setCategories y limpiar el value', () => {
    const value = 'Hola mundo';

    /*
     * 1.- Simular el input change
     * Event es para simular el event.target.value del html
     * */
    const event = { target: { value } };
    wrapper.find('input').simulate('change', event);

    /*
     * Simulacion del evento submit
     * */
    wrapper.find('form').simulate('submit', { preventDefault() {} });
    expect(setCategory).toHaveBeenCalled();
    // expect(setCategory).toHaveBeenCalledTimes(1);
    expect(setCategory).toHaveBeenCalledWith(expect.any(Function));
    expect(wrapper.find('input').prop('value')).toBe('');
  });
});
