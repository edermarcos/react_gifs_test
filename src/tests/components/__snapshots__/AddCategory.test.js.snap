// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <AddCategory /> should render correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
